src/types/impl-trait.md

```diff
@@ -111,10 +123,189 @@ fn foo() -> impl Trait {
+
+As a last usage, you can avoid binding any hidden types and just use the type-alias-impl-trait by just forwarding it elsewhere:
+
+```rust,ignore
+fn burp(x: Bar) -> Bar {
+    x
+}
+```
+
+#### Binding types
+
+You can also use type-alias-impl-trait for the type
+of local variables, constants, statics, ...
+
+```rust,ignore
+let x: Bar = 42;
+const X: Bar = 42;
+static Y: Bar = 42;
+```
+
+#### Nested in other types
+
+You can use type-alias-impl-trait in other types:
+
+```rust,ignore
+struct MyStruct {
+    bar: Bar,
+}
+```
+
+and use it just like other uses of `Bar`:
+
+```rust,ignore
+fn foo(my_struct: &MyStruct) {
+    println!("{:?}", my_struct.bar);
+}
+fn new() -> MyStruct {
+    MyStruct {
+        bar: 42
+    }
+}
+```
+
+#### Usage in trait impls
+
+Since type-alias-impl-trait can be referenced anywhere a type alias could be, this also means you can use them in `impl` blocks:
+
+```rust,ignore
+type Foo = impl Trait;
+
+impl Bar for Foo {}
+```
+
+There's a huge caveat though: now it's possible for there to be an impl for an opaque type *and* its hidden type:
+
+```rust,ignore
+type Foo = impl Trait;
+fn foo() -> Foo {}
+
+impl Bar for Foo {}
+impl Bar for () {} // ERROR conflicts with `impl Bar for Foo`
+```
+
+This check is *not* done by revealing the hidden type, but by checking whether a type could be a hidden type for that specific opaque type. So the following program is legal:
+
+```rust
+#![feature(type_alias_impl_trait)]
+trait Trait {}
+impl Trait for () {}
+type Foo = impl Trait;
+fn foo() -> Foo {}
+
+trait Bar {}
+impl Bar for Foo {}
+impl Bar for i32 {}
+```
+
+This is legal, because `i32` could not possibly be a hidden type of `Foo`, because it doesn't implement `Trait` wich is a requirement for all hypothetical hidden types of `Foo`.
+
+#### Associated types
+
+Associated types can also use `impl Trait`:
+
+```rust,ignore
+impl Deref for MyType {
+    type Target = impl Trait;
+    fn deref(&self) -> &Self::Target {
+        &self.field
+    }
+}
+```
+
+While this example is fairly artificial, the real benefit is when you have unnameable types like `async` blocks:
+
+```rust,ignore
+impl IntoFuture for MyType {
+    type Output = ();
+    type Future = impl Future<Output = ()>;
+    fn into_future(self) -> Self::Future {
+        async move {
+            // do stuff here
+        }
+    }
+}
+```
+
+This way you do not need to write burdensome `Future` impls yourself. Similarly with complex `Iterator` implementations.
+
+#### Defining scope
+
+Similar to return-position-impl-trait, you can only bind a hidden type of a type-alias-impl-trait within a specific "scope" (henceforth called "defining scope").
+The defining scope of a return-position-impl-trait is the function's body, excluding other items nested within that function's body (we may want to relax that restriction on return-position-impl-trait in the future).
```

Comment:
The note around future relaxation seems a little atypical for the reference, but maybe we can keep it.